zerver/views/message_fetch.py

```diff
@@ -936,6 +937,64 @@ def parse_anchor_value(anchor_val: Optional[str], use_first_unread_anchor: bool)
         raise JsonableError(_("Invalid anchor"))
 
 
+@has_request_variables
+def get_messages_for_ids(
+    request: HttpRequest,
+    maybe_user_profile: Union[UserProfile, AnonymousUser],
+    message_ids: List[int] = REQ(json_validator=check_list(check_int)),
+) -> HttpResponse:
+    if not maybe_user_profile.is_authenticated:
+        is_web_public_query = True
+        user_profile: Optional[UserProfile] = None
+    else:
+        is_web_public_query = False
+        assert isinstance(maybe_user_profile, UserProfile)
+        user_profile = maybe_user_profile
+        assert user_profile is not None
+
+    if is_web_public_query:
+        realm = get_valid_realm_from_request(request)
+        user_message_flags = {msg_id: ["read"] for msg_id in message_ids}
+        messages = messages_for_ids(
```

Comment:
`messages_for_ids` doesn't do any access control check; so I think this API endpoint as written would be a major security issue: Fetch any message regardless of realm given just its ID.

The fix for that is to always access Message objects using functions like `access_message`.  (This is a pattern that is intended to be used across the views codebase, for streams, messages, etc. -- always use an `access_` function that has the permission checks built in)

But we don't want to do that function in a loop.  I think `bulk_access_messages` might work for that.

For the feature we're building, we don't need the multiple message version -- we just need the single-message version.